### ISSUE TITLE
Dropdown select won't retain previous value

### ISSUE THREAD

**Issue by gjrett** [susanBuck/dwa15-spring2019#23]
I remove my form.php and uploaded the correctly named Form.php, and the name didn't change, so I must have named it form.php in an early version.

**Comment by liataun** [comment]
Perhaps related to [issue 22](https://github.com/susanBuck/dwa15-spring2019/issues/22)?

**Comment by gjrett** [comment]
Yes, that's fixed, but I'm having a hard time getting my select to hold the option chosen.  I went through the drop down example you posted, copied it completely into a test project and was able to make that work.  But when I tried applying the same code to my project, it wouldn't see "$month".  The funny thing is that the input boxes for day and year are able to see $day and $year, and I'm able to use the example logic from the notes to get them to hold their inputs.
I tried everything I could think of and checked for other solutions on the web, but no luck.  I tried changing my action so that it pointed back to the file the the select box was in, and changed the method to GET, like the example drop down, but that made no difference.  I basically spent all day Sunday trying everything I could think of, but nothing worked.
The logic that you used in the drop down example makes sense, but my problem is that I can't get the logic to see the posted selection.  I tried changing it from $month, to _$POST['month'], and just 'month', but it erred out each time with an invalid variable.
I thought I had this project done yesterday and only needed to put in the validation and get the select box to hold it's chosen value, but those two things are giving me the most trouble.  I got the validation messages from Form.php to work, but without using the html parameters, I haven't gotten the individual input boxes to go red (though I did get them to go red for some invalid inputs through CSS).  Is this enough to meet the project criteria?
I uploaded everything to github and pulled it to my server, and I changed my domain to the free one, so now it's http://p2,dwa15gjr.me.  I'd appreciate it if you could give me some guidance toward fixing these issues and let me know if i'm getting close to meeting the project criteria.
Thanks.
Best regards,Greg Retter PE PMP
    On Monday, February 25, 2019, 1:25:35 AM CST, liataun <notifications@github.com> wrote:  
 
 
Perhaps related to issue 22?

**Comment by gjrett** [comment]
Sorry, I realized I didn't tell you my github account username, it's gjrett,  
I also noticed that to keep my selection box from automatically selecting the first month, (or possibly reusing the last month selected) I had to use the multiple parameter in it's configuration. So it seems like the last month chosen stays chosen (if I don't have the multiple parameter set), but I can't use that to make the select box show the chosen month.
I apologize for the long winded dissertation, and I know you have a lot of students to deal with, but i'm hoping you can point me in the right direction.
Thanks.
Greg
    On Monday, February 25, 2019, 5:08:04 AM CST, Greg Retter <gjrett@yahoo.com> wrote:  
 
  Yes, that's fixed, but I'm having a hard time getting my select to hold the option chosen.  I went through the drop down example you posted, copied it completely into a test project and was able to make that work.  But when I tried applying the same code to my project, it wouldn't see "$month".  The funny thing is that the input boxes for day and year are able to see $day and $year, and I'm able to use the example logic from the notes to get them to hold their inputs.
I tried everything I could think of and checked for other solutions on the web, but no luck.  I tried changing my action so that it pointed back to the file the the select box was in, and changed the method to GET, like the example drop down, but that made no difference.  I basically spent all day Sunday trying everything I could think of, but nothing worked.
The logic that you used in the drop down example makes sense, but my problem is that I can't get the logic to see the posted selection.  I tried changing it from $month, to _$POST['month'], and just 'month', but it erred out each time with an invalid variable.
I thought I had this project done yesterday and only needed to put in the validation and get the select box to hold it's chosen value, but those two things are giving me the most trouble.  I got the validation messages from Form.php to work, but without using the html parameters, I haven't gotten the individual input boxes to go red (though I did get them to go red for some invalid inputs through CSS).  Is this enough to meet the project criteria?
I uploaded everything to github and pulled it to my server, and I changed my domain to the free one, so now it's http://p2,dwa15gjr.me.  I'd appreciate it if you could give me some guidance toward fixing these issues and let me know if i'm getting close to meeting the project criteria.
Thanks.
Best regards,Greg Retter PE PMP
    On Monday, February 25, 2019, 1:25:35 AM CST, liataun <notifications@github.com> wrote:  
 
 
Perhaps related to issue 22?

**Comment by susanBuck** [comment]
Hi @gjrett - 

I just cloned your project's code locally to take a look. Here are my observations:

First, you've set up your dropdown select like this:

```html
<select name='month' size='1' multiple>
```

You described above how the `multiple` was added as a workaround for your selecting issues. I don't think this is the ideal answer, and it's in conflict with the `size='1'` attribute (it does not make sense to allow multiple options but restrict your select to 1), so the first thing I did was removed both those attributes resulting in this:

```html
<select name='month'>
```

Next, looking at your options, I don't see any code that is attempting to pre-select an option:

```php
<select name='month'>
    <option value='choose'>Choose one...</option>
    <option value='January'>January</option>
    <option value='February'>February</option>
    <option value='March'>March</option>
    <option value='April'>April</option>
    <option value='May'>May</option>
    <option value='June'>June</option>
    <option value='July'>July</option>
    <option value='August'>August</option>
    <option value='September'>September</option>
    <option value='October'>October</option>
    <option value='November'>November</option>
    <option value='December'>December</option>
</select>
``` 

Skimming through your above initial post, it seems like you did at one point have the code in there, but there's no history of it in your repository, so I can't say for sure what you might have been doing right/wrong.

Given that, I'll start from scratch and work with your code, replicating the [dropdown example I provided](https://github.com/susanBuck/dwa15-php/blob/master/forms-inputs/dropdown.php). The first step here is adding the display logic of `<?php if ($month == 'x') echo 'selected' ?>` to each option:

```html
<select name='month'>
    <option value='choose' <?php if($month == null) echo 'selected'?>>Choose one...</option>
    <option value='January' <?php if ($month == 'January') echo 'selected' ?>>January</option>
    <option value='February' <?php if ($month == 'February') echo 'selected' ?>>February</option>
    <option value='March' <?php if ($month == 'March') echo 'selected' ?>>March</option>
    <option value='April' <?php if ($month == 'April') echo 'selected' ?>>April</option>
    <option value='May' <?php if ($month == 'May') echo 'selected' ?>>May</option>
    <option value='June' <?php if ($month == 'June') echo 'selected' ?>>June</option>
    <option value='July' <?php if ($month == 'July') echo 'selected' ?>>July</option>
    <option value='August' <?php if ($month == 'August') echo 'selected' ?>>August</option>
    <option value='September' <?php if ($month == 'September') echo 'selected' ?>>September</option>
    <option value='October' <?php if ($month == 'October') echo 'selected' ?>>October</option>
    <option value='November' <?php if ($month == 'November') echo 'selected' ?>>November</option>
    <option value='December' <?php if ($month == 'December') echo 'selected' ?>>December</option>
</select>
```

When the page is reloaded with these changes, we see the following notice in the dropdown:

![image](https://user-images.githubusercontent.com/237711/53359855-1a91ee80-3902-11e9-8730-be5370fea915.png)

As the notice indicates, the variable `$month` is not defined. This make sense because if we look at your logic file, `$month` only gets defined *if* there's results in the session, which there won't be on initial page load:

![image](https://user-images.githubusercontent.com/237711/53360603-f59e7b00-3903-11e9-8419-a59308f75d4a.png)


The way this is handled in the example ([dropdown-logic.php ln 3](https://github.com/susanBuck/dwa15-php/blob/master/forms-inputs/dropdown-logic.php#L3)) is to set the `$day` variable to some default (`null`) if it's not available.

```php
# Get day info from GET form request; if not present, default to `null`
$day = $_GET['day'] ?? null;
```

We could do something similar here, or we could just define `$month` to be `null` at the top, and assume it will get changed if there are results in the session. This will prevent the *undefined variable* notice.

![image](https://user-images.githubusercontent.com/237711/53360311-34800100-3903-11e9-8cce-3c150d6eceec.png)

Then, piggy-backing on this, we can update our dropdown so the first option (*Choose one...*) is what gets selected if `$month` is null:

```html
<option value='choose' <?php if($month == null) echo 'selected'?>>Choose one...</option>
```

## End result
With those changes in place, I tested the flow and confirmed the dropdown was successfully pre-selecting. [Here's a quick video run-through of that.](https://www.youtube.com/embed/s1riyx_By-E?rel=0&showinfo=0)


## Bonus tip
The display code for the select options (each month) is verbose and redundant. 
We can streamline this by defining an array of months in the logic file:

```php
$months = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'November', 'December'];
```

And then building the select options using a *for loop* that pulls from this array:

```php
<select name='month'>
    <option value='choose' <?php if ($month == null) echo 'selected' ?>>Choose one...</option>
    <?php foreach ($months as $thisMonth): ?>
        <option value='<?= $thisMonth ?>' <?php if ($month == $thisMonth) echo 'selected' ?>><?= $thisMonth ?></option>
    <?php endforeach ?>
</select>
```

**Comment by gjrett** [comment]
Thank you so much, I should have thought to pre-load the month.  The thing that threw me off was that the $day was working for the input and it wasn't pre-loaded.  
On the way to work this morning I thought of using the "if (isset( $_Session[$month]..." inside the php statement in the option, so that the $month wasn't looked at until the submit (might have worked), but you're way is much better.
I'll get that cleaned up and work on the other issues.
Thanks again.
Greg alias "@gjrett"
    On Monday, February 25, 2019, 12:54:28 PM CST, Susan Buck <notifications@github.com> wrote:  
 
 
Hi @gjrett -

I just cloned your project's code locally to take a look. Here are my observations:

First, you've set up your dropdown select like this:
<select name='month' size='1' multiple>
You described above how the multiple was added as a workaround for your selecting issues. I don't think this is the ideal answer, and it's in conflict with the size='1' attribute (it does not make sense to allow multiple options but restrict your select to 1), so the first thing I did was removed both those attributes resulting in this:
<select name='month'>
Next, looking at your options, I don't see any code that is attempting to pre-select an option:
<select name='month'>
    <option value='choose'>Choose one...</option>
    <option value='January'>January</option>
    <option value='February'>February</option>
    <option value='March'>March</option>
    <option value='April'>April</option>
    <option value='May'>May</option>
    <option value='June'>June</option>
    <option value='July'>July</option>
    <option value='August'>August</option>
    <option value='September'>September</option>
    <option value='October'>October</option>
    <option value='November'>November</option>
    <option value='December'>December</option>
</select>
Skimming through your above initial post, it seems like you did at one point have the code in there, but there's no history of it in your repository, so I can't say for sure what you might have been doing right/wrong.

Given that, I'll start from scratch and work with your code, replicating the dropdown example I provided. The first step here is adding the display logic of <?php if ($month == 'x') echo 'selected' ?> to each option:
<select name='month'>
    <option value='choose' <?php if($month == null) echo 'selected'?>>Choose one...</option>
    <option value='January' <?php if ($month == 'January') echo 'selected' ?>>January</option>
    <option value='February' <?php if ($month == 'February') echo 'selected' ?>>February</option>
    <option value='March' <?php if ($month == 'March') echo 'selected' ?>>March</option>
    <option value='April' <?php if ($month == 'April') echo 'selected' ?>>April</option>
    <option value='May' <?php if ($month == 'May') echo 'selected' ?>>May</option>
    <option value='June' <?php if ($month == 'June') echo 'selected' ?>>June</option>
    <option value='July' <?php if ($month == 'July') echo 'selected' ?>>July</option>
    <option value='August' <?php if ($month == 'August') echo 'selected' ?>>August</option>
    <option value='September' <?php if ($month == 'September') echo 'selected' ?>>September</option>
    <option value='October' <?php if ($month == 'October') echo 'selected' ?>>October</option>
    <option value='November' <?php if ($month == 'November') echo 'selected' ?>>November</option>
    <option value='December' <?php if ($month == 'December') echo 'selected' ?>>December</option>
</select>
When the page is reloaded with these changes, we see the following notice in the dropdown:



As the notice indicates, the variable $month is not defined. This make sense because if we look at your logic file, $month only gets defined if there's results in the session, which there won't be on initial page load:



The way this is handled in the example (dropdown-logic.php ln 3) is to set the $day variable to some default (null) if it's not available.
# Get day info from GET form request; if not present, default to `null`
$day = $_GET['day'] ?? null;
We could do something similar here, or we could just define $month to be null at the top, and assume it will get changed if there are results in the session. This will prevent the undefined variable notice.



Then, piggy-backing on this, we can update our dropdown so the first option (Choose one...) is what gets selected if $month is null:
<option value='choose' <?php if($month == null) echo 'selected'?>>Choose one...</option>
End result

With those changes in place, I tested the flow and confirmed the dropdown was successfully pre-selecting. Here's a quick video run-through of that.

Bonus tip

The display code for the select options (each month) is verbose and redundant.
We can streamline this by defining an array of months in the logic file:
$months = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'November', 'December'];
And then building the select options using a for loop that pulls from this array:
<select name='month'>
    <option value='choose' <?php if ($month == null) echo 'selected' ?>>Choose one...</option>
    <?php foreach ($months as $thisMonth): ?>
        <option value='<?= $thisMonth ?>' <?php if ($month == $thisMonth) echo 'selected' ?>><?= $thisMonth ?></option>
    <?php endforeach ?>
</select>

**Comment by liataun** [comment]
This was helpful for me. I ran into a similar issue with maintaining `select` values. I worked out a solution based off of the text field examples I had already built and came up with a similar set of pieces as above, if with slightly different logic. The *Bonus Tip* will be useful if I need to expand on my options in the future. (Currently my selection is between two, but has potential for several more.)

**Comment by gjrett** [comment]
Susan,
i did the assignment and the changes to my Foobooks project to match yours, by following the videos and notes.  I also made changes to my P3 project but tried to load two different packages to provide Zodiac info for the input date (recall my project gives the day of the week for an input date).
I troubleshot the package installs and thought I had one installed, but couldn't get it to work.  I ended up downloading last weeks save from Github, then tried to rebuild it to at least get it back to last weeks working state, but I keep getting a 500 error when I try to run it.  I've copied a vendor folder into the project and ran the composer update, but I still get the 500 error, and I can't think of what to do next.
I work from 5:00am (that's when I leave for work) until 3:30pm (when I usually get off) and on Tues. and Weds. I teach from 5:00pm to about 9:00pm.  So with travel time, I won't have much time to get my project working properly.  Is the work I've done on the assignment and my Foobooks (which has been pushed to my Github account) enough to complete this weeks work, or do I need to incorporate your changes into my project and get it working on my droplet?
Thanks.
Best regards,Greg Retter
